File: FlutterProjects/Practice/Backend with JWT- v4 Upgrading/src/main/java/com/mrahmed/HRandPayrollManagementSystem/service/AttendanceService.java
package com.mrahmed.HRandPayrollManagementSystem.service;

import com.mrahmed.HRandPayrollManagementSystem.entity.Attendance;
import com.mrahmed.HRandPayrollManagementSystem.entity.User;
import com.mrahmed.HRandPayrollManagementSystem.repository.AttendanceRepository;
import com.mrahmed.HRandPayrollManagementSystem.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class AttendanceService {
    @Autowired
    private AttendanceRepository attendanceRepository;

    @Autowired
    private UserRepository userRepository;

    // Method to get overtime for a user within a date range with pagination
    public Page<Attendance> getOvertimeForUser(Long userId, LocalDate startDate, LocalDate endDate, Pageable pageable) {
        Page<Attendance> attendances = attendanceRepository.findAttendancesByUserIdAndDateRange(userId, startDate, endDate, pageable);
        return attendances.map(this::filterOvertime);
    }

    // Check if the attendance record contains overtime
    private Attendance filterOvertime(Attendance attendance) {
        if (attendance.getClockInTime() == null || attendance.getClockOutTime() == null) {
            return null; // Skip records with missing times
        }
        Duration duration = Duration.between(attendance.getClockInTime(), attendance.getClockOutTime());
        long hours = duration.toHours();
        return hours > 8 ? attendance : null; // Return only overtime records
    }

    // Check-in method
    public Attendance checkIn(long userId) {
        try {
            User user = userRepository.findById(userId)
                    .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));
            if (attendanceRepository.countByUserIdAndDate(userId, LocalDate.now()) > 0) {
                throw new RuntimeException("User has already checked in today.");
            }
            Attendance attendance = new Attendance();
            attendance.setDate(LocalDate.now());
            attendance.setClockInTime(LocalDateTime.now());
            attendance.setUser(user);
            return attendanceRepository.save(attendance);
        } catch (Exception e) {
            throw new RuntimeException("Error occurred while checking in: " + e.getMessage(), e);
        }
    }

    // Check-out method
    public Attendance checkOut(long userId) {
        try {
            Attendance attendance = attendanceRepository.findLastAttendanceForUser(userId, LocalDate.now())
                    .orElseThrow(() -> new RuntimeException("No check-in found for today's check-out"));
            if (attendance.getClockOutTime() != null) {
                throw new RuntimeException("User has already checked out today.");
            }
            attendance.setClockOutTime(LocalDateTime.now());
            return attendanceRepository.save(attendance);
        } catch (Exception e) {
            throw new RuntimeException("Error occurred while checking out: " + e.getMessage(), e);
        }
    }

    // Fetch today's attendance with pagination
    public Page<Attendance> getTodayAttendances(Pageable pageable) {
        return attendanceRepository.findAttendancesForToday(LocalDate.now(), pageable);
    }

    // Fetch all attendance records with pagination
    public Page<Attendance> getAllAttendances(Pageable pageable) {
        return attendanceRepository.findAll(pageable);
    }

    // Find attendance by ID
    public Attendance findAttendanceById(long id) {
        return attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found with id: " + id));
    }

    // Get users' attendance count in a date range with pagination
    public Map<User, Long> getUsersAttendanceInRange(LocalDate startDate, LocalDate endDate, Pageable pageable) {
        Page<Attendance> attendances = attendanceRepository.findAttendancesInRange(startDate, endDate, pageable);
        return attendances.getContent().stream()
                .collect(Collectors.groupingBy(Attendance::getUser, Collectors.counting()));
    }

    // Get all attendance records for a specific user with pagination
    public Page<Attendance> getAttendanceByUserId(Long id, Pageable pageable) {
        return attendanceRepository.findAllByUserId(id, pageable);
    }

    // Find users without attendance today with pagination
    public Page<User> findUsersWithoutAttendanceToday(Pageable pageable) {
        return attendanceRepository.findUsersWithoutAttendanceForToday(LocalDate.now(), pageable);
    }

    // Fetch peak attendance day with pagination
    public Page<Object[]> getPeakAttendanceDay(Pageable pageable) {
        return attendanceRepository.findPeakAttendanceDay(pageable);
    }

    // Fetch peak attendance month with pagination
    public Page<Object[]> getPeakAttendanceMonth(Pageable pageable) {
        return attendanceRepository.findPeakAttendanceMonth(pageable);
    }

    // Get late check-ins with pagination
    public Page<Attendance> getLateCheckIns(String lateTime, LocalDate startDate, LocalDate endDate, Pageable pageable) {
        return attendanceRepository.findLateCheckIns(lateTime, startDate, endDate, pageable);
    }

    // Search attendance records by userName with pagination
    public Page<Attendance> getAttendancesByUserNamePart(String name, Pageable pageable) {
        return attendanceRepository.findAttendancesByUserNamePart(name, pageable);
    }

    // Filter attendance by user role and date range with pagination
    public Page<Attendance> getAttendanceByRole(String role, Pageable pageable) {
        return attendanceRepository.findAttendanceByRole(role, pageable);
    }

    // Get today's attendance by user ID with pagination
    public Page<Attendance> getTodayAttendanceByUserId(long userId, Pageable pageable) {
        return attendanceRepository.findTodayAttendanceByUserId(userId, pageable);
    }

}

File: FlutterProjects/Practice/Flutter Backup/hr_and_pms - v1 Attendance Added/lib/features/attendance/service/AttendanceService.dart
import 'dart:convert';
import 'package:hr_and_pms/features/attendance/model/AttendanceModel.dart';
import 'package:hr_and_pms/features/user/model/User.dart';
import 'package:hr_and_pms/wrapper_model/PaginatedAttendance.dart';
import 'package:http/http.dart' as http;

class AttendanceService {
  final String baseUrl = 'localhost:8080';
  final String apiPath = '/api/attendance';

  // Method for checking in
  Future<Attendance> checkIn(int userId) async {
    final response = await http.post(
      Uri.http(baseUrl, '$apiPath/checkin'),
      body: json.encode({'userId': userId}),
      headers: {'Content-Type': 'application/json'},
    );

    if (response.statusCode == 201) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to check in: ${response.body}');
    }
  }

  // Method for checking out
  Future<Attendance> checkOut(int userId) async {
    final response = await http.put(
      Uri.http(baseUrl, '$apiPath/checkout'),
      body: json.encode({'userId': userId}),
      headers: {'Content-Type': 'application/json'},
    );

    if (response.statusCode == 200) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to check out: ${response.body}');
    }
  }

  // Method to get overtime for a user within a date range with pagination
  Future<PaginatedAttendance> getOvertimeByUser(int userId, DateTime startDate, DateTime endDate, {int page = 0, int size = 10}) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/overtime/$userId', {
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
        'page': '$page',
        'size': '$size',
      }),
    );

    if (response.statusCode == 200) {
      return PaginatedAttendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load overtime records: ${response.body}');
    }
  }

  // Method to get today's attendance records
  Future<List<Attendance>> getTodayAttendance() async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/today'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load today’s attendance: ${response.body}');
    }
  }


  // Method to retrieve all attendances with pagination
  Future<PaginatedAttendance> getAllAttendances({int page = 0, int size = 10}) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/', {'page': '$page', 'size': '$size'}),
    );

    if (response.statusCode == 200) {
      return PaginatedAttendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load all attendances: ${response.body}');
    }
  }


  // Method to find attendance by ID
  Future<Attendance> findAttendanceById(int id) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/find/$id'),
    );

    if (response.statusCode == 200) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Attendance not found: ${response.body}');
    }
  }

  // Method to get attendance records for a specific user
  Future<List<Attendance>> getAttendancesByUserId(int userId, {int page = 0, int size = 10}) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/user/$userId/attendances', {'page': '$page', 'size': '$size'}),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load user attendance records: ${response.body}');
    }
  }

  // Method to get attendance counts in a range
  Future<Map<String, dynamic>> getUsersWithAttendanceInRange(DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/attendanceRange', {
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
      }),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load attendance count in range: ${response.body}');
    }
  }

  // Method to get users without attendance today
  Future<List<User>> getUsersWithoutAttendanceToday() async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/usersWithoutAttendanceToday'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => User.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load users without attendance today: ${response.body}');
    }
  }

  // Method to get peak attendance day
  Future<List<Object>> getPeakAttendanceDay() async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/peakAttendanceDay'),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load peak attendance day: ${response.body}');
    }
  }

  // Method to get peak attendance month
  Future<List<Object>> getPeakAttendanceMonth() async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/peakAttendanceMonth'),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load peak attendance month: ${response.body}');
    }
  }

  // Method to get late check-ins
  Future<List<Attendance>> getLateCheckIns(String lateTime, DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/lateCheckIns', {
        'lateTime': lateTime,
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
      }),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load late check-ins: ${response.body}');
    }
  }

  // Method to search attendance by username
  Future<List<Attendance>> getAttendancesByUserNamePart(String name, {int page = 0, int size = 10}) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/searchByName', {'name': name, 'page': '$page', 'size': '$size'}),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load attendances by user name: ${response.body}');
    }
  }

  // Method to get attendance by role
  Future<List<Attendance>> getAttendanceByRole(String role, {int page = 0, int size = 10}) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/roleAttendance', {'role': role, 'page': '$page', 'size': '$size'}),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load attendances by role: ${response.body}');
    }
  }

  // Method to get today's attendance by user ID
  Future<List<Attendance>> getTodayAttendanceByUserId(int userId) async {
    final response = await http.get(
      Uri.http(baseUrl, '$apiPath/todayAttendance/$userId'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load today’s attendance by user ID: ${response.body}');
    }
  }
}
